rust/connlib/tunnel/src/iface_handler.rs

```diff
@@ -168,31 +168,23 @@ where
         self: &Arc<Self>,
         iface_config: Arc<IfaceConfig>,
         device_io: DeviceIo,
-    ) {
+    ) -> std::io::Result<()> {
         let device_writer = device_io.clone();
+        let mut src = [0u8; MAX_UDP_SIZE];
+        let mut dst = [0u8; MAX_UDP_SIZE];
         loop {
-            let mut src = [0u8; MAX_UDP_SIZE];
-            let mut dst = [0u8; MAX_UDP_SIZE];
-            let res = {
-                // TODO: We should check here if what we read is a whole packet
-                // there's no docs on tun device on when a whole packet is read, is it \n or another thing?
-                // found some comments saying that a single read syscall represents a single packet but no docs on that
-                // See https://stackoverflow.com/questions/18461365/how-to-read-packet-by-packet-from-linux-tun-tap
-                match device_io.read(&mut src[..iface_config.mtu()]).await {
-                    Ok(res) => res,
-                    Err(e) => {
-                        tracing::error!(error = ?e, from = "iface", action = "read");
-                        let _ = self.callbacks.on_error(&e.into());
-                        continue;
-                    }
-                }
-            };
+            let res = device_io.read(&mut src[..iface_config.mtu()]).await?;
+            if res == 0 {
+                return Ok(());
+            }
 
             tracing::trace!(target: "wire", action = "read", bytes = res, from = "iface");
```

Comment:
Perhaps this should be _before_ the early exit on `0` to also capture that in the logs?